src/test/PdfTest/PdfControllerUnitTests.java

```diff
@@ -0,0 +1,733 @@
+    }
+
+    choices = new JSONArray();
+    choices.put("Choice1");
+    choices.put("Choice2");
+    choices.put("Choice3");
+    formAnswers.put("Combobox", choices);
+    correctFieldValues.put("Combobox", "[Choice1, Choice2, Choice3]");
+
+    pdfInput = new File("src/test/resources/testpdf.pdf");
+    pdfDocument = PDDocument.load(pdfInput);
+    PdfController.fillFields(pdfDocument, formAnswers);
+    fieldValues = PdfController.getFieldValues(pdfDocument);
+
+    // We test that all the fields in correctFields have the right value
+    for (String key : correctFieldValues.keySet()) {
+      System.out.println(key + " " + fieldValues.getString(key));
+      Assert.assertEquals(correctFieldValues.getString(key), fieldValues.getString(key));
+    }
+    pdfDocument.close();
+  }
+
+  @Test
+  public void fillFieldsSignatureFieldTest() throws IOException {
+    File pdfInput = new File("src/test/resources/testpdf.pdf");
+    InputStream imageStream = new FileInputStream("src/test/resources/first-love.png"); // PNG
+    PDDocument pdfDocument = PDDocument.load(pdfInput);
+
+    PDVisibleSignDesigner visibleSignDesigner = new PDVisibleSignDesigner(imageStream);
+    visibleSignDesigner.zoom(0);
+    PDVisibleSigProperties visibleSigProperties =
+        new PDVisibleSigProperties()
+            .visualSignEnabled(true)
+            .setPdVisibleSignature(visibleSignDesigner);
+    visibleSigProperties.buildSignature();
+
+    PDSignature signature = new PDSignature();
+    signature.setFilter(PDSignature.FILTER_ADOBE_PPKLITE);
+    signature.setSubFilter(PDSignature.SUBFILTER_ADBE_PKCS7_DETACHED);
+    signature.setName("Example Name1");
+    signature.setLocation("Philadelphia, PA");
+    signature.setReason("Application");
+    signature.setSignDate(Calendar.getInstance());
+
+    SignatureOptions signatureOptions = new SignatureOptions();
+    signatureOptions.setVisualSignature(visibleSigProperties.getVisibleSignature());
+    PDSignatureField signatureField =
+        (PDSignatureField) pdfDocument.getDocumentCatalog().getAcroForm().getField("Signature");
+    signatureField.setValue(signature);
+    pdfDocument.addSignature(signature, signatureOptions);
+
+    PDSignature signatureDocument = signatureField.getSignature();
+    Assert.assertEquals(signature.getFilter(), signatureDocument.getFilter());
+    Assert.assertEquals(signature.getSubFilter(), signature.getSubFilter());
+    Assert.assertEquals(signature.getName(), signatureDocument.getName());
+    Assert.assertEquals(signature.getLocation(), signatureDocument.getLocation());
+    Assert.assertEquals(signature.getReason(), signatureDocument.getReason());
+    Assert.assertEquals(signature.getSignDate(), signatureDocument.getSignDate());
+    Assert.assertEquals(
+        signature.getCOSObject().getItem(COSName.CONTENTS),
+        signatureDocument.getCOSObject().getItem((COSName.CONTENTS)));
+    pdfDocument.save("src/test/resources/test_out_signature.pdf");
```

Comment:
if we are saving this file, are we deleting it anywhere or what

Reply:
Done